File: hestiaGO/hestiaTESTING/renderer.go
// Copyright 2022 "Holloway" Chew, Kean Ho <kean.ho.chew@zoralab.com>
// Copyright 2022 ZORALab Enterprise <tech@zoralab.com>
//
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//                  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.

package hestiaTESTING

// string
const (
	header_STRING = "\n" +
		"╔═══════════╗" + "\n" +
		"║TEST REPORT║" + "\n" +
		"╚═══════════╝" + "\n"

	title_ID_STRING = DATA_LABEL_ID + "\t\t: "
	end_ID_STRING   = "\n"

	title_NAME_STRING = DATA_LABEL_NAME + "\t\t: "
	end_NAME_STRING   = "\n"

	title_VERDICT_STRING = DATA_LABEL_VERDICT + "\t\t: "
	end_VERDICT_STRING   = "\n"

	title_DESCRIPTION_STRING = DATA_LABEL_DESCRIPTION + "\t:\n"
	end_DESCRIPTION_STRING   = "\n\n\n"

	title_SWITCHES_STRING = DATA_LABEL_SWITCHES + "\t:\n"
	open_SWITCH_STRING    = "["
	close_SWITCH_STRING   = "]\n"
	end_SWITCH_STRING     = "\n"
	end_SWITCHES_STRING   = "\n\n"

	title_LOG_STRING = DATA_LABEL_LOG + "\t\t:" + "\n"
	open_LOG_STRING  = "["
	close_LOG_STRING = "]\n"
	end_LOG_STRING   = "\n\n"

	footer_STRING = "═══[ END ]═══\n\n\n"
)

// toml
const (
	escapeQUOTE_TOML = "'''"

	header_TOML = "[" + DATA_LABEL_GROUP + "]" + "\n"

	title_ID_TOML = DATA_LABEL_ID + " = "
	end_ID_TOML   = "\n"

	title_VERDICT_TOML = DATA_LABEL_VERDICT + " = '"
	end_VERDICT_TOML   = "'\n"

	title_NAME_TOML = DATA_LABEL_NAME + " = '''\n"
	end_NAME_TOML   = "\n'''\n"

	title_DESCRIPTION_TOML = DATA_LABEL_DESCRIPTION + " = '''\n"
	end_DESCRIPTION_TOML   = "\n'''\n"

	empty_SWITCHES_TOML = DATA_LABEL_SWITCHES + " = []\n"
	title_SWITCH_TOML   = "\n[[" + DATA_LABEL_GROUP + "." + DATA_LABEL_SWITCHES + "]]" +
		"\n" +
		DATA_LABEL_VALUE + " = " + escapeQUOTE_TOML + "\n"
	end_SWITCH_TOML = "\n" + escapeQUOTE_TOML + "\n"

	empty_LOG_WITH_EMPTY_SWITCHES_TOML = DATA_LABEL_LOG + " = []\n"
	empty_LOG_TOML                     = "\n[" + DATA_LABEL_GROUP + "]\n" +
		empty_LOG_WITH_EMPTY_SWITCHES_TOML
	title_LOG_TOML = "\n[[" + DATA_LABEL_GROUP + "." + DATA_LABEL_LOG + "]]\n" +
		DATA_LABEL_VALUE + " = " + escapeQUOTE_TOML + "\n"
	end_LOG_TOML = "\n" + escapeQUOTE_TOML + "\n"
)

func _checkBeforeRender(s *Scenario, name string) {
	if s == nil {
		panic("calling hestiaTESTING.To" + name + " without providing Scenario!")
	}

	s.Init()
}

File: hestiaRUST/hestia_testing/renderer.rs
// Copyright 2022 "Holloway" Chew, Kean Ho <kean.ho.chew@zoralab.com>
// Copyright 2022 ZORALab Enterprise <tech@zoralab.com>
//
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//                  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.
use crate::hestia_testing::data;
use crate::hestia_testing::scenario;

pub fn to_string(s: &scenario::Scenario) -> String {
	// render switches
	let mut switches = String::from("");
	for (i, v) in s.switches.iter().enumerate() {
		switches.push_str(&format!("[{}]\t{}\n", i, v));
	}

	// render log
	let mut logs = String::from("");
	for (i, v) in s.logs.iter().enumerate() {
		logs.push_str(&format!("[{}]\n{}\n", i, v));
	}

	// render output
	return format!(
		"
╔═══════════╗
║TEST REPORT║
╚═══════════╝
{label_id}		: {id}
{label_verdict}		: {verdict}
{label_name}		: {name}
{label_description}	:
{description}


{label_switches}\t	:
{switches}


{label_log}		:
{logs}
═══[ END ]═══
",
		label_id = data::LABEL_ID,
		id = s.id,
		label_verdict = data::LABEL_VERDICT,
		verdict = data::interpret(scenario::conclusion(s)),
		label_name = data::LABEL_NAME,
		name = s.name,
		label_description = data::LABEL_DESCRIPTION,
		description = s.description,
		label_switches = data::LABEL_SWITCHES,
		switches = switches,
		label_log = data::LABEL_LOG,
		logs = logs,
	);
}

pub fn to_toml(s: &scenario::Scenario) -> String {
	// render switches
	let mut switches = String::from("");
	for (_i, v) in s.switches.iter().enumerate() {
		switches.push_str(&format!(
			"
[[{label_group}.{label_switches}]]
{label_value} = '''
{switch}
'''
",
			label_group = data::LABEL_GROUP,
			label_switches = data::LABEL_SWITCHES,
			label_value = data::LABEL_VALUE,
			switch = v,
		));
	}

	if switches.chars().count() == 0 {
		switches.push_str(&format!(
			"{label_switches} = []",
			label_switches = data::LABEL_SWITCHES,
		));
	}

	// render log
	let mut logs = String::from("");
	for (_i, v) in s.logs.iter().enumerate() {
		switches.push_str(&format!(
			"
[[{label_group}.{label_log}]]
{label_value} = '''
{entry}
'''
",
			label_group = data::LABEL_GROUP,
			label_log = data::LABEL_LOG,
			label_value = data::LABEL_VALUE,
			entry = v,
		));
	}

	if logs.chars().count() == 0 {
		logs.push_str(&format!("{label_log} = []", label_log = data::LABEL_LOG,));
	}

	// render output
	return format!(
		"
[{label_group}]
{label_id} = {id}
{label_verdict} = '{verdict}'
{label_name} = '''
{name}
'''
{label_description} = '''
{description}
'''
{switches}
{logs}
",
		label_group = data::LABEL_GROUP,
		label_id = data::LABEL_ID,
		id = s.id,
		label_verdict = data::LABEL_VERDICT,
		verdict = data::interpret(scenario::conclusion(s)),
		label_name = data::LABEL_NAME,
		name = s.name,
		label_description = data::LABEL_DESCRIPTION,
		description = s.description,
		switches = switches,
		logs = logs,
	);
}
